Add metric health schema to Entities

pavan-traceable opened the pull request:
**Add baseline for metrics for entities**

## Description
Baseline, lower bound and upper bound values for entity metrics and used gateway service implementation. 

<!--
- **on a feature**: describe the feature and how this change fits in it, e.g. this PR makes kafka message.max.bytes configurable to better support batching
- **on a refactor**: describe why this is better than previous situation e.g. this PR changes logic for retry on healthchecks to avoid false positives
- **on a bugfix**: link relevant information about the bug (github issue or slack thread) and how this change solves it e.g. this change fixes #99999 by adding a lock on read/write to avoid data races.
-->


### Testing
Please describe the tests that you ran to verify your changes. Please summarize what did you test and what needs to be tested e.g. deployed and tested helm chart locally. 

### Checklist:
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] Any dependent changes have been merged and published in downstream modules

### Documentation
Make sure that you have documented corresponding changes in this repository or [hypertrace docs repo](https://github.com/hypertrace/hypertrace-docs-website) if required.

<!--
Include __important__ links regarding the implementation of this PR.
This usually includes and RFC or an aggregation of issues and/or individual conversations that helped put this solution together. This helps ensure there is a good aggregation of resources regarding the implementation.
-->

aaron-steinfeld commented on hypertrace-graphql-gateway-service-metric-utils/src/main/java/org/hypertrace/graphql/utils/metrics/gateway/BaselineConvertedAggregationContainer.java:
```diff
@@ -0,0 +1,59 @@
+import static org.hypertrace.core.graphql.attributes.AttributeModelMetricAggregationType.DISTINCT_COUNT;
+
+@AllArgsConstructor
+public class BaselineConvertedAggregationContainer implements BaselineMetricAggregationContainer {
```
implementations should not be public

pavan-traceable replied:
ok

aaron-steinfeld commented on hypertrace-graphql-metric-schema/src/main/java/org/hypertrace/graphql/metric/schema/Health.java:
```diff
@@ -0,0 +1,17 @@
+    String HEALTH_MESSAGE = "message";
+
+    @GraphQLName(HEALTH_VALUE)
+    @GraphQLField
```
Can we make these non nullable

aaron-steinfeld commented on hypertrace-graphql-metric-schema/src/main/java/org/hypertrace/graphql/metric/schema/BaselineMetricInterval.java:
```diff
@@ -0,0 +1,25 @@
+
+    @GraphQLName(BASELINE_AGGREGATION_VALUE)
+    @GraphQLField
+    BaselineAggregation baseline();
```
What is the baseline here? Isn't the baseline value inside the aggregations?

aaron-steinfeld commented on hypertrace-graphql-metric-schema/src/main/java/org/hypertrace/graphql/metric/schema/BaselineMetricAggregationContainer.java:
```diff
@@ -0,0 +1,57 @@
+
+import static org.hypertrace.graphql.metric.schema.MetricAggregationContainer.*;
+
+public interface BaselineMetricAggregationContainer {
```
Can we make MetricAggregationContainer an interface to make sure the two variants of it stay in sync?

aaron-steinfeld commented on hypertrace-graphql-metric-schema/src/main/java/org/hypertrace/graphql/metric/schema/BaselineMetricAggregation.java:
```diff
@@ -0,0 +1,15 @@
+import graphql.annotations.annotationTypes.GraphQLName;
+
+@GraphQLName(BaselineMetricAggregation.BASELINE_METRIC_AGGREGATION_TYPE_NAME)
+public interface BaselineMetricAggregation extends MetricAggregation {
```
Naming throughout - this name sounds like it's a baseline value, when in reality it's both an aggregation and it's baseline. Could we rename it `BaselinedMetricAggregation` or propose a new name?

aaron-steinfeld commented on hypertrace-graphql-metric-schema/src/main/java/org/hypertrace/graphql/metric/schema/BaselineAggregation.java:
```diff
@@ -0,0 +1,24 @@
+import graphql.annotations.annotationTypes.GraphQLField;
+import graphql.annotations.annotationTypes.GraphQLName;
+
+public interface BaselineAggregation extends MetricAggregation {
```
A couple naming nits:
Could we call this one `MetricBaselineAggregation`

and lowerValue/upperValue lowerBound/upperBound (or similar, I think "value" doesn't convey that it's a range of expectation).

tim-mwangi commented on hypertrace-graphql-gateway-service-metric-utils/src/main/java/org/hypertrace/graphql/utils/metrics/gateway/MetricContainerMapConverter.java:
```diff
@@ -71,13 +70,13 @@
 
     ConvertedMetricContainer(
-        Map<MetricLookupMapKey, MetricAggregation> metricAggregationMap,
+        Map<MetricLookupMapKey, BaselineMetricAggregation> metricAggregationMap,
```
BaselineMetricAggregation implies something totally different from the existing MetricAggregation. Are you sure you want to do a replace/rename instead of an augmentation to the existing users of MetricAggregation with a different object of the type BaselineMetricAggregation?

aaron-steinfeld commented on hypertrace-graphql-gateway-service-metric-utils/src/main/java/org/hypertrace/graphql/utils/metrics/gateway/MetricContainerMapConverter.java:
```diff
@@ -87,5 +86,10 @@
+
+    @Override
+    public List<BaselineMetricInterval> baselineSeries(int size, TimeUnit units) {
+      return null;
```
Never return null, please have this return an empty list with a comment that it's not implemented.

aaron-steinfeld commented on hypertrace-graphql-gateway-service-metric-utils/src/main/java/org/hypertrace/graphql/utils/metrics/gateway/MetricContainerMapConverter.java:
```diff
@@ -71,13 +70,13 @@
   }
 
-  private static class ConvertedMetricContainer extends ConvertedAggregationContainer
+  private static class ConvertedMetricContainer extends BaselineConvertedAggregationContainer
```
make sure implementation and interface names stay in sync

pavan-traceable replied:
done

aaron-steinfeld commented on hypertrace-graphql-gateway-service-metric-utils/src/main/java/org/hypertrace/graphql/utils/metrics/gateway/MetricContainerMapConverter.java:
```diff
@@ -21,9 +22,7 @@
-import org.hypertrace.graphql.metric.schema.MetricAggregation;
-import org.hypertrace.graphql.metric.schema.MetricContainer;
-import org.hypertrace.graphql.metric.schema.MetricInterval;
+import org.hypertrace.graphql.metric.schema.*;
```
no wildcard imports, make sure you're using the google java style plugin.

aaron-steinfeld commented on hypertrace-graphql-gateway-service-metric-utils/src/main/java/org/hypertrace/graphql/utils/metrics/gateway/MetricAggregationMapConverter.java:
```diff
@@ -27,16 +27,16 @@
-      List<MetricAggregationRequest> aggregationRequests,
-      Map<String, AggregatedMetricValue> resultMap) {
+  public Single<Map<MetricLookupMapKey, BaselineMetricAggregation>> convert(
+          List<MetricAggregationRequest> aggregationRequests,
```
bad indentation - google java style plugin please

aaron-steinfeld commented on hypertrace-graphql-gateway-service-metric-utils/src/main/java/org/hypertrace/graphql/utils/metrics/gateway/MetricAggregationConverter.java:
```diff
@@ -27,7 +29,12 @@
+
+    @Override
+    public BaselineAggregation baseline() {
+      return null;
```
not implemented comment, impl/interface name match

pavan-traceable commented:
@tim-mwangi  @aaron-steinfeld thanks for the quick review.  This patch is not complete it needs refactoring i should have updated description well, i have just updated just to verify with you guys regarding schema changes. Since generics is not supported in graphql so want to check whether schema changes fine or not.

aaron-steinfeld commented:
> @tim-mwangi @aaron-steinfeld thanks for the quick review. This patch is not complete it needs refactoring i should have updated description well, i have just updated just to verify with you guys regarding schema changes. Since generics is not supported in graphql so want to check whether schema changes fine or not.

Just to clarify, generics can be used (and should be, for enforcing keeping objects in sync) - they just need to be reified in child interfaces so the reflection code can find the appropriate types.

pavan-traceable commented:
@aaron-steinfeld  Please review the implementation and API changes. Some minor things pending will address along with review comments. Not able to find placeholder to calculate Health to metrics, please suggest the right place.

aaron-steinfeld commented on hypertrace-graphql-entity-schema/src/main/java/org/hypertrace/graphql/entity/dao/GatewayServiceEntityDao.java:
```diff
@@ -48,9 +52,10 @@
-            serverRequest -> this.makeRequest(request.resultSetRequest().context(), serverRequest))
-        .flatMap(serverResponse -> this.entityConverter.convert(request, serverResponse));
+        .flatMap(serverRequest ->  this.makeRequest(request.resultSetRequest().context(), serverRequest)
+        .flatMap(serverResponse -> baselineProvider.buildRequest(serverRequest, serverResponse, request)
```
this indentation is off, hiding the nesting. Please update, and break into multiple functions if the nesting seems confusing.

aaron-steinfeld commented on hypertrace-graphql-entity-schema/src/main/java/org/hypertrace/graphql/entity/dao/GatewayServiceEntityDao.java:
```diff
@@ -24,6 +25,7 @@
   private final GraphQlGrpcContextBuilder grpcContextBuilder;
   private final GatewayServiceEntityRequestBuilder requestBuilder;
   private final GatewayServiceEntityConverter entityConverter;
+  private final BaselineProvider baselineProvider;
```
How is BaselineProvider conceptually different than the abstract DAO interfaces we have? Do we need another concept or could this become `BaselineDao`? Provider is also already a term used in this project any any providers should implement `javax.inject.Provider`.

aaron-steinfeld commented on hypertrace-graphql-entity-schema/src/main/java/org/hypertrace/graphql/entity/health/BaselineProvider.java:
```diff
@@ -0,0 +1,15 @@
+import org.hypertrace.graphql.entity.request.EntityRequest;
+
+public interface BaselineProvider {
+    Single<BaselineEntitiesRequest> buildRequest(EntitiesRequest entitiesRequest, EntitiesResponse entitiesResponse, EntityRequest request);
```
Any benefit of separating these? build + make requests when they're implemented together?

aaron-steinfeld replied:
Separate methods is fine and good from an implementation standpoint - but not from an API. Is there any reason to expose an api that needs two steps just to break up the implementation?

aaron-steinfeld commented on hypertrace-graphql-gateway-service-metric-utils/src/main/java/org/hypertrace/graphql/utils/metrics/gateway/BaselineConvertedAggregationContainer.java:
```diff
@@ -0,0 +1,63 @@
+import static org.hypertrace.core.graphql.attributes.AttributeModelMetricAggregationType.*;
+import static org.hypertrace.core.graphql.attributes.AttributeModelMetricAggregationType.DISTINCT_COUNT;
+
+@Accessors(fluent = true)
```
What does this generate given that the whole interface appears to be implemented below?

aaron-steinfeld commented on hypertrace-graphql-gateway-service-metric-utils/src/main/java/org/hypertrace/graphql/utils/metrics/gateway/MetricBaselineAggregationConverter.java:
```diff
@@ -0,0 +1,51 @@
+    implements Converter<Optional<Baseline>, MetricBaselineAggregation> {
+
+  @Override
+  public Single<MetricBaselineAggregation> convert(Optional<Baseline> baseline) {
```
Generally we don't want to accept optionals as an arg - the caller shouldn't call the function if it doesn't have a baseline to convert. Please remove the default valueing happening below and null out the whole baseline object if we don't have a value to convert.

aaron-steinfeld commented on hypertrace-graphql-metric-schema/src/main/java/org/hypertrace/graphql/metric/request/DefaultMetricAggregationRequestBuilder.java:
```diff
@@ -70,12 +72,18 @@
+          AttributeModel attribute,
+          AttributeModelMetricAggregationType aggregationType,
+          List<Object> arguments) {
+    return new DefaultMetricAggregationRequest(attribute, aggregationType, arguments, Collections.EMPTY_LIST);
```
```suggestion
    return new DefaultMetricAggregationRequest(attribute, aggregationType, arguments, Collections.emptyList());
```

`EMPTY_LIST` returns a raw list - I can't think of any times `emptyList()` wouldn't be preferable.

aaron-steinfeld replied:
This is also the issue I was referring to on the interface - any calls to this api would silently omit baselines on the response.

aaron-steinfeld commented on hypertrace-graphql-metric-schema/src/main/java/org/hypertrace/graphql/metric/request/DefaultMetricAggregationRequestBuilder.java:
```diff
@@ -70,12 +72,18 @@
+  public MetricAggregationRequest build(
+          AttributeModel attribute,
+          AttributeModelMetricAggregationType aggregationType,
+          List<Object> arguments, List<SelectedField> selections) {
```
formatter.

aaron-steinfeld commented on hypertrace-graphql-metric-schema/src/main/java/org/hypertrace/graphql/metric/request/MetricAggregationRequestBuilder.java:
```diff
@@ -22,4 +22,10 @@ MetricAggregationRequest build(
+          AttributeModel attribute,
+          AttributeModelMetricAggregationType aggregationType,
+          List<Object> arguments,
+          List<SelectedField> selections);
```
Which selections are these meant to be? the sub-selections of the aggregation? If they're needed, to resolve baselines (which likely they are), how do we make the other APIs compatible without silently breaking. In other words, doesn't this belong as a modification of the existing API?

aaron-steinfeld replied:
This is the problem that I'm trying to point out though - existing calls won't break according to the compiler, but if they contain baselines those baselines won't be fetched so it will functionally be broken. We should modify the existing api and force ourselves to explicitly make the decision about baselines on the caller side.

aaron-steinfeld commented on hypertrace-graphql-metric-schema/src/main/java/org/hypertrace/graphql/metric/request/DefaultMetricAggregationRequestBuilder.java:
```diff
@@ -157,11 +167,20 @@ private Integer percentileSize(Map<String, Object> arguments) {
     AttributeModel attribute;
     AttributeModelMetricAggregationType aggregation;
     List<Object> arguments;
+    List<SelectedField> selectedFields;
```
please don't put graphql framework constructs in the generated request - the whole purpose of the request builders is to parse the graphql objects into our own pojo. Instead, do this logic in the builder and set the appropriate value directly.

aaron-steinfeld commented on hypertrace-graphql-metric-schema/src/main/java/org/hypertrace/graphql/metric/request/DefaultMetricAggregationRequestBuilder.java:
```diff
@@ -157,11 +167,20 @@ private Integer percentileSize(Map<String, Object> arguments) {
+      if (selectedFields.size() > 0) {
+        return new Baseline();
+      }
+      return null;
```
never null - this likely goes away based on my comment on the baseline class - seems like a boolean.

aaron-steinfeld commented on hypertrace-graphql-metric-schema/src/main/java/org/hypertrace/graphql/metric/request/DefaultMetricAggregationRequestBuilder.java:
```diff
@@ -85,7 +93,9 @@ private MetricAggregationRequest requestForAggregationField(
-        this.getArgumentsForAggregation(aggregationType, field.getArguments()));
+        this.getArgumentsForAggregation(aggregationType, field.getArguments()),
+            this.selectionFinder.findSelections(field.getSelectionSet(),
+                    SelectionQuery.namedChild(BASELINE_AGGREGATION_VALUE)).collect(Collectors.toList()));
```
combined with other comments, you could get a boolean out of this with `.findAny().isPresent()`

aaron-steinfeld commented on hypertrace-graphql-entity-schema/src/main/java/org/hypertrace/graphql/entity/health/GatewayBaselineProviderImpl.java:
```diff
@@ -0,0 +1,146 @@
+    return baselineMap;
+  }
+
+  private List<FunctionExpression> getFunctionExpression(
```
nit: naming, this returns a list not a single expression

aaron-steinfeld commented on hypertrace-graphql-entity-schema/src/main/java/org/hypertrace/graphql/entity/health/GatewayBaselineProviderImpl.java:
```diff
@@ -0,0 +1,146 @@
+  }
+
+  private Map<String, Baseline> getBaselineRequestMap(EntityRequest request) {
+    Map<String, Baseline> baselineMap = new HashMap<>();
```
As a general practice, always try to avoid mutable collections - easy source of bugs. What you're doing here can be easily written with a `stream/flatmap/collect`

aaron-steinfeld commented on hypertrace-graphql-entity-schema/src/main/java/org/hypertrace/graphql/entity/health/GatewayBaselineProviderImpl.java:
```diff
@@ -0,0 +1,146 @@
+  private Map<String, Baseline> getBaselineRequestMap(EntityRequest request) {
+    Map<String, Baseline> baselineMap = new HashMap<>();
+    List<MetricRequest> metricRequests = request.metricRequests();
+    if (metricRequests.size() > 0) {
```
unnecessary if

aaron-steinfeld commented on hypertrace-graphql-entity-schema/src/main/java/org/hypertrace/graphql/entity/health/GatewayBaselineProviderImpl.java:
```diff
@@ -0,0 +1,146 @@
+      for (MetricRequest metricRequest : metricRequests) {
+        baselineMap.putAll(
+            metricRequest.aggregationRequests().stream()
+                .filter(x -> x.baseline() != null)
```
nit: even in lambdas, real variable names help readability

aaron-steinfeld replied:
if baseline() returned a boolean as suggested elsewhere, this coded would also simplify. Are there any places we're using it beyond a null check?

aaron-steinfeld commented on hypertrace-graphql-entity-schema/src/main/java/org/hypertrace/graphql/entity/health/GatewayBaselineProviderImpl.java:
```diff
@@ -0,0 +1,146 @@
+        baselineMap.putAll(
+            metricRequest.aggregationRequests().stream()
+                .filter(x -> x.baseline() != null)
+                .collect(Collectors.toMap(x -> x.alias(), y -> y.baseline())));
```
nit: could be moved to method references

aaron-steinfeld commented on hypertrace-graphql-entity-schema/src/main/java/org/hypertrace/graphql/entity/dao/GatewayServiceEntityConverter.java:
```diff
@@ -74,15 +81,30 @@
+  private Optional<BaselineEntity> getBaselineEntity(Map<String, BaselineEntity> baselineEntityMap,
+                                           org.hypertrace.gateway.service.v1.entity.Entity entity) {
+    if (!baselineEntityMap.isEmpty() && baselineEntityMap.containsKey(entity.getId())) {
+      return Optional.of(baselineEntityMap.get(entity.getId()));
```
nit `Optional.ofNullable` would replace the if here since a get for a missing id would just return null.

aaron-steinfeld commented on hypertrace-graphql-entity-schema/src/main/java/org/hypertrace/graphql/entity/dao/GatewayServiceEntityConverter.java:
```diff
@@ -74,15 +81,30 @@
+  }
+
+  private Map<String, BaselineEntity> getEntityMap(BaselineEntitiesResponse baselineResponse) {
+    List<BaselineEntity> entityList = baselineResponse.getBaselineEntityList();
```
nit: unnecessary variable

aaron-steinfeld commented on hypertrace-graphql-gateway-service-metric-utils/src/main/java/org/hypertrace/graphql/utils/metrics/gateway/MetricAggregationContainerMapConverter.java:
```diff
@@ -39,15 +41,15 @@
         .flatMap(
-            metricRequests -> this.aggregationMapConverter.convert(metricRequests, metricResponses))
-        .map(ConvertedAggregationContainer::new)
+            metricRequests -> this.aggregationMapConverter.convert(metricRequests, metricResponses, Collections.EMPTY_MAP))
```
emptyMap()

aaron-steinfeld commented on hypertrace-graphql-gateway-service-metric-utils/src/main/java/org/hypertrace/graphql/utils/metrics/gateway/MetricAggregationContainerMapConverter.java:
```diff
@@ -39,15 +41,15 @@
-            metricRequests -> this.aggregationMapConverter.convert(metricRequests, metricResponses))
-        .map(ConvertedAggregationContainer::new)
+            metricRequests -> this.aggregationMapConverter.convert(metricRequests, metricResponses, Collections.EMPTY_MAP))
+        .map(aggMap -> new BaselineConvertedAggregationContainer((Map<MetricLookupMapKey, BaselinedMetricAggregation>) aggMap))
```
Why is the cast needed? can we fix the types and remove the cast, then switch to `BaselineConvertedAggregationContainer::new`

aaron-steinfeld commented on hypertrace-graphql-gateway-service-metric-utils/src/main/java/org/hypertrace/graphql/utils/metrics/gateway/MetricAggregationMapConverter.java:
```diff
@@ -1,45 +1,95 @@
+            this.baselineAggregationConverter.convert(
+                Optional.ofNullable(baselineMap.get(aggregationRequest.alias()))),
+            (metricAggregation, baselinedMetricAggregation) ->
+                new BaselinedMetricAggregation() {
```
Can we make real (inner or otherwise) classes and remove the inline anonymous class? Makes the code much harder to follow

aaron-steinfeld commented on hypertrace-graphql-gateway-service-metric-utils/src/main/java/org/hypertrace/graphql/utils/metrics/gateway/MetricAggregationMapConverter.java:
```diff
@@ -1,45 +1,95 @@
+    return zip(
+            this.aggregationConverter.convert(resultMap.get(aggregationRequest.alias()).getValue()),
+            this.baselineAggregationConverter.convert(
+                Optional.ofNullable(baselineMap.get(aggregationRequest.alias()))),
```
optional as an arg always puts up red flags. Unless they really make the code harder to follow, optionals should really only be used as returns.

aaron-steinfeld commented on hypertrace-graphql-gateway-service-metric-utils/src/main/java/org/hypertrace/graphql/utils/metrics/gateway/MetricAggregationMapConverter.java:
```diff
@@ -1,45 +1,95 @@
+
+                  @Override
+                  public MetricBaselineAggregation baseline() {
+                    return new MetricBaselineAggregation() {
```
Here too - shouldn't this be what the converter returns?

aaron-steinfeld commented on hypertrace-graphql-gateway-service-metric-utils/src/main/java/org/hypertrace/graphql/utils/metrics/gateway/BaselineIntervalListContainer.java:
```diff
@@ -0,0 +1,180 @@
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+class BaselineIntervalListContainer
```
make sure all the class names match the classes they produce - here BaselineMetricInterval rather than BaselineInterval

aaron-steinfeld commented on hypertrace-graphql-gateway-service-metric-utils/src/main/java/org/hypertrace/graphql/utils/metrics/gateway/BaselineIntervalListContainer.java:
```diff
@@ -0,0 +1,180 @@
+  private final MetricBaselineAggregationConverter baselineAggregationConverter;
+
+  @Inject
+  BaselineIntervalListContainer(MetricBaselineAggregationConverter baselineAggregationConverter) {
```
this looks like mostly copy paste from the existing interval list converter - how much of the logic could be shared? I expect most of it could go in a base class that is just parameterized with the specific impementation builders.

aaron-steinfeld commented on hypertrace-graphql-gateway-service-metric-utils/src/main/java/org/hypertrace/graphql/utils/metrics/gateway/BaselineIntervalListContainer.java:
```diff
@@ -0,0 +1,180 @@
+      this.endTime = intervalTimeRange.getEndTime();
+    }
+
+    // TODO Fix this
```
Assume this is due to the null below?

aaron-steinfeld commented on hypertrace-graphql-gateway-service-metric-utils/src/main/java/org/hypertrace/graphql/utils/metrics/gateway/BaselineIntervalListContainer.java:
```diff
@@ -0,0 +1,180 @@
+
+      @Override
+      public Health health() {
+        // TODO modify this
```
what's the plan on health?

aaron-steinfeld commented on hypertrace-graphql-gateway-service-metric-utils/src/main/java/org/hypertrace/graphql/utils/metrics/gateway/BaselineMetricSeriesMapConverter.java:
```diff
@@ -0,0 +1,61 @@
+  public Single<Map<Duration, List<BaselineMetricInterval>>> convert(
+      List<MetricSeriesRequest> metricSeriesRequests,
+      Map<String, BaselineMetricSeries> baselineMetricSeriesMap) {
+    return Observable.fromIterable(metricSeriesRequests)
```
Same as intervallist, this looks like copy paste that could be shared.

pavan-traceable replied:
As i said earlier will do refactoring in next iteration since it already became big patch.

aaron-steinfeld commented:
> @aaron-steinfeld Please review the implementation and API changes. Some minor things pending will address along with review comments. Not able to find placeholder to calculate Health to metrics, please suggest the right place.

Lots of comments, many of them nit/style. Can we please first format this code with google style (in 1 standalone commit), then merge in main (again standalone), then fix any compilation issues, since it's not even running in CI and I'm fairly sure it won't compile as is.

As to health, my recommendation would be to omit it (from the schema as well as the implementation). We need to be far more iterative with this, and we can easily add health in a later round - the scope of this is already very large and difficult to review.

pavan-traceable commented on .circleci/config.yml:
```diff
@@ -158,7 +158,7 @@ workflows:
           requires:
             - build
-            - snyk-scan
+            - synk-scan
```
ok

pavan-traceable commented on hypertrace-graphql-entity-schema/src/main/java/org/hypertrace/graphql/entity/dao/GatewayServiceEntityRequestBuilder.java:
```diff
@@ -24,7 +24,7 @@
 import org.hypertrace.graphql.metric.schema.argument.AggregatableOrderArgument;
 
-class GatewayServiceEntityRequestBuilder {
+public class GatewayServiceEntityRequestBuilder {
```
ok

jcchavezs commented:
> Lots of comments, many of them nit/style. Can we please first format this code with google style (in 1 standalone commit), then merge in main (again standalone), then fix any compilation issues, since it's not even running in CI and I'm fairly sure it won't compile as is.

My 2p in here. Could we add a GH action checking this by default? That way we avoid unrelated changes in PRs due to formatting. cc @JBAhire 

> As to health, my recommendation would be to omit it (from the schema as well as the implementation). We need to be far more iterative with this, and we can easily add health in a later round - the scope of this is already very large and difficult to review.

Yes please, this PR is already big. Maybe it is easier to break it up in small PRs? Otherwise reviewing this is almost impossible and it is easy to make mistakes. Please consider either merge smaller PRs against `main` or another feature branch as they are easier to review.

aaron-steinfeld commented:
> My 2p in here. Could we add a GH action checking this by default? That way we avoid unrelated changes in PRs due to formatting. cc @JBAhire

+1 !

jcchavezs commented:
Related to formating on CI: https://github.com/hypertrace/document-store/pull/24

aaron-steinfeld commented on hypertrace-graphql-metric-schema/src/main/java/org/hypertrace/graphql/metric/request/MetricAggregationRequest.java:
```diff
@@ -13,4 +13,6 @@
 
   List<Object> arguments();
+
+  Baseline baseline();
```
I'm not sure we have a use case for that, but we don't support fetching those independently anyway, do we?

aaron-steinfeld commented on hypertrace-graphql-gateway-service-metric-utils/src/main/java/org/hypertrace/graphql/utils/metrics/gateway/MetricAggregationMapConverter.java:
```diff
@@ -57,39 +57,51 @@
+
+    private final MetricBaselineAggregation metricBaselineAggregation;
+
+    public MetricBaselineAggregationImpl(MetricBaselineAggregation metricBaselineAggregation) {
```
What is the purpose of this class? It looks like its only constructor is a clone constructor

pavan-traceable replied:
Removed, not required

aaron-steinfeld commented on hypertrace-graphql-entity-schema/src/main/java/org/hypertrace/graphql/entity/dao/GatewayServiceEntityDao.java:
```diff
@@ -52,10 +52,21 @@
-        .flatMap(baselineRequest -> baselineProvider.makeRequest(request.resultSetRequest().context(), baselineRequest))
-        .flatMap(baselineResponse -> this.entityConverter.convert(request, serverResponse, baselineResponse))));
+        .flatMap(
+            serverRequest ->
```
The formatting here is meant to expose the complexity of following this logic, which I think is now apparent. Between the comments about reducing the api of `BaselineDao`, and packaging together the responses to avoid the triconverter (not sure what the issues you ran into with that were), I think we'd see this simplify.

pavan-traceable replied:
Reduced the api for BaselineDao.

pavan-traceable replied:
I am not sure why tri-converter simplify this method,  it just reduces one argument here. Tri-converter is only used for 2 classes, i don't think it added much complexity.

aaron-steinfeld commented on hypertrace-graphql-entity-schema/src/main/java/org/hypertrace/graphql/entity/health/GatewayBaselineDao.java:
```diff
@@ -32,20 +33,17 @@
 
 @Singleton
-public class GatewayBaselineProviderImpl implements BaselineProvider {
+public class GatewayBaselineDao implements BaselineDao {
```
package private

pavan-traceable replied:
done

aaron-steinfeld commented on hypertrace-graphql-entity-schema/src/main/java/org/hypertrace/graphql/entity/health/GatewayBaselineDao.java:
```diff
@@ -57,47 +55,53 @@
-      EntitiesRequest entitiesRequest, Map<String, Baseline> requestMap) {
-    List<Expression> selectionlist = entitiesRequest.getSelectionList();
+  private List<FunctionExpression> getFunctionExpressionList(
+      EntitiesRequest entitiesRequest, Map<String, Boolean> requestMap) {
```
This requestmap is only being used to detect presence - a Set would be a better fit.

pavan-traceable replied:
right, fixed it

aaron-steinfeld commented on hypertrace-graphql-entity-schema/src/main/java/org/hypertrace/graphql/entity/health/GatewayBaselineDao.java:
```diff
@@ -57,47 +55,53 @@
-      }
-    }
+  private Map<String, Boolean> getBaselineRequestMap(EntityRequest request) {
+    Map<String, Boolean> baselineMap =
+        request.metricRequests().stream()
+            .flatMap(
+                metricRequest ->
+                    metricRequest.aggregationRequests().stream()
+                        .filter(MetricAggregationRequest::baseline))
+            .collect(
+                Collectors.toMap(
+                    MetricAggregationRequest::alias, MetricAggregationRequest::baseline));
+    baselineMap.putAll(
+        request.metricRequests().stream()
+            .flatMap(
+                metricRequest ->
+                    metricRequest.seriesRequests().stream()
+                        .filter(
+                            timeSeriesRequest -> timeSeriesRequest.aggregationRequest().baseline()))
+            .collect(
+                Collectors.toMap(
+                    MetricSeriesRequest::alias,
+                    seriesRequest -> seriesRequest.aggregationRequest().baseline())));
```
IMO, avoiding nesting makes this easier to follow (also switched to set based on other comment)

```suggestion
          Stream<String> aggregationAliases = request.metricRequests().stream()
            .map(MetricRequest::aggregationRequests)
            .flatMap(Collections::stream)
            .filter(MetricAggregationRequest::baseline)
            .map(MetricAggregationRequest::alias);
            
          Stream<String> seriesAliases = request.metricRequests().stream()
            .map(MetricRequest::seriesRequests)
            .flatMap(Collections::stream)
            .filter(seriesRequests -> seriesRequests.aggregationRequest().baseline())
            .map(MetricAggregationRequest::alias);
            
          return Streams.concat(aggregationAliases, seriesAliases)
            .collect(Collectors.toUnmodifiableSet())
```

pavan-traceable replied:
Thanks will change accordingly

pavan-traceable commented:
Compilation issues are because it depends on gateway-service which is not merged yet.

JBAhire commented:
> Related to formating on CI: [hypertrace/document-store#24](https://github.com/hypertrace/document-store/pull/24)

Will incorporate this. Thanks @jcchavezs :)

pavan-traceable commented:
@jcchavezs We can divide this PR into API and implementation, but API part is small compared to implementation. From next time i will try to push small PR's, if possible

pavan-traceable commented on hypertrace-graphql-gateway-service-metric-utils/src/main/java/org/hypertrace/graphql/utils/metrics/gateway/MetricBaselineAggregationConverter.java:
```diff
@@ -0,0 +1,62 @@
+    }
+  }
+
+  static class MetricBaselineAggregationDefaultInstance implements MetricBaselineAggregation {
```
@aaron-steinfeld  Please suggest if this can be improved further.

pavan-traceable commented:
@aaron-steinfeld  I have updated all comments. Please review.

aaron-steinfeld commented on hypertrace-graphql-metric-schema/src/main/java/org/hypertrace/graphql/metric/schema/BaselineMetricInterval.java:
```diff
@@ -0,0 +1,21 @@
+import java.time.Instant;
+
+@GraphQLName(BaselineMetricInterval.BASELINE_METRIC_INTERVAL)
+public interface BaselineMetricInterval extends BaselineMetricAggregationContainer {
```
A `BaselineMetricAggregationContainer` contains both baseline and non-baseline data (see my comment on its naming) - which means both would show up on the series. This doesn't match the API we had discussed, where we settled on splitting the series (which we've _also_ done here).

pavan-traceable replied:
BaselinedMetricAggregationContainer doesn't contains both baseline and non baseline data. It contains 3 fields for each aggregation function, Value, lowerBound and UpperBound.

pavan-traceable replied:
<img width="1611" alt="Screenshot 2021-01-08 at 1 29 25 PM" src="https://user-images.githubusercontent.com/73101820/103989340-9410ee00-51b5-11eb-83a8-62dd21a69cce.png">
Attached screenshot of schema

pavan-traceable replied:
You are right, fixed the schema

pavan-traceable replied:
<img width="1347" alt="Screenshot 2021-01-08 at 2 19 09 PM" src="https://user-images.githubusercontent.com/73101820/103994059-7e52f700-51bc-11eb-9f03-899c9b764c33.png">
New schema

aaron-steinfeld commented on hypertrace-graphql-gateway-service-metric-utils/src/main/java/org/hypertrace/graphql/utils/metrics/gateway/BaselineMetricIntervalListContainer.java:
```diff
@@ -0,0 +1,158 @@
+    private static Map<MetricLookupMapKey, BaselinedMetricAggregation> getBaselineAggMap(
+        Map<MetricLookupMapKey, MetricBaselineAggregation> metricAggregationMap) {
+      Map<MetricLookupMapKey, BaselinedMetricAggregation> baselineMap = new HashMap<>();
+      metricAggregationMap.forEach(
```
Processing should happen in the converter, not the data container - but this highlights the issue I discussed on the interfaces - we've added an extra layer of nesting here, where we'd have:
```
  baselineSeries: [
     {
        sum: {
           value: null // This is meaningless, it's harded coded to null below
            baseline: { // This is a redundant layer of nesting
               value: <value>
            }
        }
     }
  ]

pavan-traceable replied:
Fixed this and removed all processing here.

pavan-traceable commented on hypertrace-graphql-entity-schema/src/main/java/org/hypertrace/graphql/entity/dao/GatewayServiceEntityConverter.java:
```diff
@@ -51,21 +57,22 @@
                 this.convertEntity(
                     request,
                     entity,
+                    getBaselineEntity(getEntityMap(baselineResponse), entity),
```
done

pavan-traceable commented on hypertrace-graphql-metric-schema/src/main/java/org/hypertrace/graphql/metric/request/MetricSeriesRequestBuilder.java:
```diff
@@ -42,6 +43,16 @@
         .flatMap(field -> this.flattenAggregationsForSeries(attribute, field));
   }
 
+  Observable<MetricSeriesRequest> buildBaselineSeries(
```
Done

pavan-traceable commented on hypertrace-graphql-gateway-service-metric-utils/src/main/java/org/hypertrace/graphql/utils/metrics/gateway/BaselineMetricIntervalListContainer.java:
```diff
@@ -0,0 +1,158 @@
+import org.hypertrace.graphql.metric.schema.BaselineMetricInterval;
+import org.hypertrace.graphql.metric.schema.BaselinedMetricAggregation;
+import org.hypertrace.graphql.metric.schema.MetricBaselineAggregation;
+
```
Yes did that sometime back, even though i formatted entire code using google formatter, it didnt deleted that gap. Anyway i am deleting myself.

aaron-steinfeld replied:
Hmm - did you use the intellij format config or the plugin? I know the format file isn't able to capture the full style, the plugin might have shortcomings too - I use both, but that feels excessive. We'll revisit this with the push to enforce this in CI (and maybe add a gradle task)

pavan-traceable commented on hypertrace-graphql-metric-schema/src/main/java/org/hypertrace/graphql/metric/schema/BaselineMetricAggregationContainer.java:
```diff
@@ -0,0 +1,64 @@
+import org.hypertrace.graphql.metric.schema.argument.MetricAggregationAvgRateUnitsArgument;
+import org.hypertrace.graphql.metric.schema.argument.MetricAggregationPercentileSizeArgument;
+
+public interface BaselineMetricAggregationContainer extends MetricAggregationContainer {
```
Done

pavan-traceable commented on hypertrace-graphql-entity-schema/src/main/java/org/hypertrace/graphql/entity/health/BaselineDao.java:
```diff
@@ -0,0 +1,17 @@
+import org.hypertrace.graphql.entity.request.EntityRequest;
+
+public interface BaselineDao {
+  Single<BaselineEntitiesResponse> buildAndSubmitRequest(
```
Done

pavan-traceable commented on hypertrace-graphql-gateway-service-metric-utils/src/main/java/org/hypertrace/graphql/utils/metrics/gateway/MetricAggregationMapConverter.java:
```diff
@@ -1,45 +1,81 @@
         Map<String, AggregatedMetricValue>,
-        Map<MetricLookupMapKey, MetricAggregation>> {
+        Map<String, Baseline>,
+        Map<MetricLookupMapKey, BaselinedMetricAggregation>> {
```
done

pavan-traceable commented on hypertrace-graphql-gateway-service-metric-utils/src/main/java/org/hypertrace/graphql/utils/metrics/gateway/MetricAggregationMapConverter.java:
```diff
@@ -1,45 +1,81 @@
+
+    @Override
+    public Double value() {
+      return metricAggregation.value();
```
done.

pavan-traceable commented on hypertrace-graphql-gateway-service-metric-utils/src/main/java/org/hypertrace/graphql/utils/metrics/gateway/MetricContainerMapConverter.java:
```diff
@@ -8,84 +8,126 @@
+
+  private Map<String, BaselineMetricSeries> getBaselineMetricSeriesMap(
+      BaselineEntity baselineEntity) {
+    if (baselineEntity == null) {
```
Passed default instances across the patch.

pavan-traceable commented:
@aaron-steinfeld  I have addressed all your comments, only thing pending is removing nesting inline in GatewayServiceEntityDao which can be  done  in next iteration. UI team is already blocked on this. Lets push this, but before that we should merge https://github.com/hypertrace/hypertrace-core-graphql/pull/43

aaron-steinfeld commented on hypertrace-graphql-gateway-service-metric-utils/src/main/java/org/hypertrace/graphql/utils/metrics/gateway/MetricAggregationMapConverter.java:
```diff
@@ -54,28 +56,34 @@
     @Override
-    public Double value() {
-      return metricAggregation.value();
+    public MetricBaselineAggregation baseline() {
```
These two methods shouldn't be needed with the lombok annotations above

pavan-traceable commented:
I have updated core-graphql still it is showing conflicts.

aaron-steinfeld commented on hypertrace-graphql-gateway-service-metric-utils/src/main/java/org/hypertrace/graphql/utils/metrics/gateway/MetricBaselineAggregationConverter.java:
```diff
@@ -13,50 +14,22 @@
-    public Double value() {
-      return null;
-    }
+    static final MetricBaselineAggregation EMPTY =
```
So this would end up with null values rather than a null baseline object - we can go with that and revisit

aaron-steinfeld commented:
> I have updated core-graphql still it is showing conflicts.

You'll need to rebase this repo off main, then update core-graphql. Once that's done, I think we're good at a good iteration point.

pavan-traceable commented:
@aaron-steinfeld  All checks have passed.
